### ISSUE TITLE
Feature Request: Auto-select License

### ISSUE THREAD

**Issue by xcpttrips** [Linj1k/fab_extended#1]
As the extension is named Fab Extended and not Fab Wishlist, I am assuminfg more features are planned.
Here's a suggestion (and I'll drop another idea as a seperate issue):

The new licensing options on Fab are annoying.
It would be great if the extension would auto select the best available option, i.e.:
1. Free Professional
2. Free Personal
3. Personal
4. Professional

[Always selecting Personal (or Professional as a choice in settings) would most likely suffice, even though I've seen assets that were set to Free Professional with a price for the Personal (though those might have been in error).]

**Comment by Linj1k** [comment]
Hello, Exactly, I was planning to add functionality, hence the name,
I think it's possible, making changes on Fab is rather complex because it's made in react which means that the components of the page are replaced each time changes are made,

I'll look at your suggestions when I have a bit of time to spare

Thank you

**Comment by Linj1k** [comment]
Hello,
I've added your feature, but we'll have to wait for the Chrome Web Store to validate it.

I haven't added a settings menu, as I don't find it useful for the time it would take to do so.
- Auto-select License (Automatically selects the personal license, if this is not possible, the first license will be selected)

**Comment by xcpttrips** [comment]
I played around a bit with the license auto-selector one and have it working for now (had to repair it after Epic made changes, which btw. currently also removed all traces of the Favorites functionality from Fab Extended for me from fab.com).

Here's my code for reference (never wrote a browser extension before, so no idea about best practices, etc., but after some trial and error it does what I wanted it to do with the license):

ContentScript.js:

```
// Constants for delays
const DROPDOWN_OPEN_DELAY = 300; // Adjusted for dropdown rendering
const RETRY_INTERVAL = 300; // Retry interval

// Function to locate and click the dropdown menu
function openDropdown() {
    const dropdownButton = document.querySelector(
        '#root > div.A89t3u91 > main > div > aside > div > div.fabkit-Surface-root.fabkit-Surface--emphasis-background-elevated-low-transparent.fabkit-scale--gutterX-spacing-8.fabkit-scale--gutterY-spacing-8.fabkit-Stack-root.fabkit-scale--gapX-spacing-4.fabkit-scale--gapY-spacing-4.fabkit-Stack--column.ny_JIXbd > div.fabkit-FormField-root.fabkit-FormField-root.fabkit-FormField--hideLabel > div.fabkit-FormField-container > button'
    );

    if (!dropdownButton) {
        console.log("[DEBUG] Dropdown menu not found.");
        return null;
    }

    dropdownButton.click();
    console.log("[DEBUG] Dropdown clicked.");
    return dropdownButton;
}

// Function to check if the dropdown popup is visible
function isDropdownVisible() {
    const dropdownPopup = document.querySelector('aside.fabkit-Dropdown-container');
    if (dropdownPopup && getComputedStyle(dropdownPopup).display !== 'none') {
        console.log("[DEBUG] Dropdown popup is now visible.");
        return dropdownPopup;
    }
    console.log("[DEBUG] Dropdown popup is not visible.");
    return null;
}

// Function to handle the dropdown popup
// Function to filter and handle dropdown options
function handleDropdownPopup() {
    const dropdown = isDropdownVisible();
    if (dropdown) {
        const options = Array.from(dropdown.querySelectorAll('div, span, li'));
        console.log(`[DEBUG] Found ${options.length} options in the dropdown.`);

        // Filter to exclude combined or irrelevant options
        const priorityOptions = options.filter(option => {
            const text = option.textContent.trim();
            return (text.includes("Free") || text.includes("Personal")) &&
                !text.includes("results available") && // Exclude combined options
                !text.includes("Professional");         // Avoid combined Professional options
        });

        if (priorityOptions.length > 0) {
            console.log(`[DEBUG] Priority option found: ${priorityOptions[0].textContent}`);
            priorityOptions[0].click();
            console.log("[DEBUG] Priority option clicked.");
        } else {
            console.log("[DEBUG] No valid 'Free' or 'Personal' option found.");
        }
    } else {
        console.log("[DEBUG] Dropdown popup is still not visible.");
        // Retry with a small delay
        setTimeout(handleDropdownPopup, 200);  // Retry after 200ms if dropdown isn't visible
    }
}

// Ensure the MutationObserver watches for any necessary changes
const observer = new MutationObserver(() => {
    console.log("[DEBUG] MutationObserver triggered.");
    handleDropdownPopup();
});

observer.observe(document.body, {
    childList: true,
    subtree: true
});


// Retry logic to ensure dropdown handling
function retryHandleDropdown() {
    const intervalId = setInterval(() => {
        console.log("[DEBUG] Retrying dropdown handling...");
        const dropdown = openDropdown();
        if (dropdown) {
            setTimeout(() => {
                const visiblePopup = isDropdownVisible();
                if (visiblePopup) {
                    handleDropdownPopup();
                    clearInterval(intervalId);
                }
            }, DROPDOWN_OPEN_DELAY);
        }
    }, RETRY_INTERVAL);
}

// Start the retry logic
retryHandleDropdown();
console.log("[DEBUG] Script initialized and running.");
```

manifest.json:

```
{
    "manifest_version": 3,
    "name": "FAB.com License Selector",
    "version": "1.0",
    "description": "Automatically selects the 'Free Professional' or 'Free Personal' or 'Personal' license option on fab.com listings.",
    "permissions": [ "scripting", "activeTab" ],
    "content_scripts": [
        {
            "matches": [ "https://www.fab.com/listings/*" ],

            "js": [ "contentScript.js" ],
            "run_at": "document_idle"
        }
    ],
    "icons": {
        "48": "icons/icon48.png",
        "128": "icons/icon128.png"
    }
}

```

**Comment by Linj1k** [comment]
I saw the problem and solved it,
We'll have to find a long-term solution.

I'd never done an extension before this one
I'll let you take a look at how I approached the feature: https://github.com/Linj1k/fab_extended/commit/b64543ec941e63f02338f408bb0a91514c51653b

**Comment by xcpttrips** [comment]
Looks like the feature works reliably.

**Comment by xcpttrips** [comment]
I found one case where the current implementation can fail.

Steps to reproduce: Open product page for asset owned with a Professional license (e.g. aquired via Humble Bundle or Monthly Featured Content)

Observed Result: Extension selects Personal License which causes the asset to show as if not yet owned.

Note: this isn't reproducable 100% of the time and seems to depend on random timing (i.e. fab.com and the extension both change the state and the order of execution can differ between tries).

Expected Result: License selection would need to take Owned status into consideration.

**Comment by Linj1k** [comment]
Can you please give me a link to a product that has this problem?

**Comment by xcpttrips** [comment]
This could be tricky, as my license on any particular asset might be different from yours even if we both own them.
You would need to have an asset where you are certain you own the Professional license (it will show in the selection of the popup when its on screen, not sure if we can check the license version someplace else).

I could capture a video or get the console log if the extension saves anything to there.

Looks like the Sponsored (Free For The Month) content that was claimed on the old marketplace, transferred to Fab as a personal license, so those won't do. 

However, the 3 that are currently featured, are Free in both types of licenses (though if you already claimed them with a personal license, those won't glitch either): 

https://www.fab.com/blade/c949409a-866e-4e74-8d1b-208cd257e899?context=homepage 

If you had a secondary Fab account, where those weren't yet claimed, you could easily test with these 3.
Select Professional license and add to account, then enable extension and reload those pages (several times if it doesn't glitch right away).
 
Other than that, if you own any UE specific Humble Bundles, I could check if it applies to any of those if I own them as well.

**Comment by Linj1k** [comment]
I think I've corrected the problem,
I've changed the function whose personal license is detected,

I've also made several global modifications that should make the extension a lot nicer.

**Comment by xcpttrips** [comment]
Is the change in the live version or still pending?
In Version: 0.1.6 the license selection still favors a free personal license over an owned professional one.

After page load:
![image](https://github.com/user-attachments/assets/ef791266-cdb9-4149-ac11-b6ffd9ae01fc)


With popup visible (manually clicked):
![image](https://github.com/user-attachments/assets/17f50159-5d6c-489b-9165-f70a4ad2713e)

**Comment by Linj1k** [comment]
You mean that if the license you want is already selected, the selection should choose the other one?

**Comment by xcpttrips** [comment]
Nevermind, the issue was, as I suspected, due to the live version trailing behind (the one from the store was 0.1.6).
I now loaded 0.1.8 from Github and it is working correctly!

**Comment by Linj1k** [comment]
Perfect, yes the chrome web store takes a while to validate updates

**Comment by xcpttrips** [comment]
License selection doesn't work with this asset page:
https://www.fab.com/listings/96b9fca5-9bcd-4456-923c-3077ea4ac044?quickBuyOfferId=ac6bb4a1e1734dd3a4aac282de5ba86f

I own a Professional license but the extension forces the Personal one (just like I reported with v0.1.6 in this thread).
When the page loads, it flashes "You own this" for a split second, then it selects the Personal license with its price.
My current version is 0.1.8 from Github.

Nothing seems to differ from the others I've tested (e.g. the 3 that are free for a limited time - those  still work fine, tested today).
